### PR TITLE
[LegalizeTypes] Handle non byte-sized elt types when splitting INSERT/EXTRACT_VECTOR_ELT

### PR DESCRIPTION
**Pull request by bjope**
DAGTypeLegalizer::SplitVecRes_INSERT_VECTOR_ELT and
DAGTypeLegalizer::SplitVecRes_EXTRACT_VECTOR_ELT did not handle
non byte-sized elements properly. In fact, it only dealt with
elements smaller than 8 bits (as well as byte-sized elements).

This patch generalizes the support for non byte-sized element by
always widening the the vector elements to next "round integer type"
(a power of 2 bit size). This should make sure that we can access a
single element via a simple byte-addressed scalar load/store.

Also removing a suspicious CustomLowerNode call from
SplitVecRes_INSERT_VECTOR_ELT. Considering that it did not reset
the Lo/Hi out arguments before the return I think that
DAGTypeLegalizer::SplitVectorResult could be fooled into registering
the input vector as being the result. This should however not have
caused any problems since DAGTypeLegalizer::SplitVectorResult is
doing the same CustomLowerNode call, making the code removed by
this patch redundant.

### CODE REVIEW

**Comment by efriedma-quic** [comment]
Is the nearest byte really what we want here?  We can widen to any type we want, and on most targets, storing an `<N x i32>` is going to be a lot more efficient than storing an `<N x i24>`.

Instead of expanding immediately, should we just construct an EXTRACT_VECTOR_ELT with the promoted vector type, and legalize that?  That might allow custom legalization to kick in in certain cases.

**Comment by arsenm** [comment]
> > Is the nearest byte really what we want here? We can widen to any type we want, and on most targets, storing an `<N x i32>` is going to be a lot more efficient than storing an `<N x i24>`.
> 
> It is a bit hard to understand what is best for every target (unless perhaps scanning for some legal type, if such a type exists). Some reasons for not picking a larger element type would be:

Using the next legal element type would be a reasonable default. Maybe you want next legal store memory type 

> 
> * Our downstream target can store things like i24, so using getRoundIntegerType would both introduce an unneeded extend, but also wasting stack space. 

As in i24 is a legal type, or truncating store i32->i24 is legal?

**Comment by arsenm** [comment]
> So it seems to make more sense in-tree (given by what ValueTypes.td currently looks like) to make use of getRoundIntegerType here. Or alternatively we should do something slightly more complicated by trying to find a legal type for ISD::STORE. But I don't know if there already exist some helper for that or if we would need to invent it.
> 
> I'll try to update this PR to use getRoundIntegerType instead of the getStoreSizedIntegerVectorElementType. And then we will deal with it downstream, if we want special handling for types like i24.
>

getTypeToTransformTo?

**Comment by efriedma-quic** [comment]
> A target could just promote the input vector elements itself if that is feasible when doing CustomLowerNode the first time. 

For non-MVT types, I don't think there's any way for a target to request custom lowering.  At least, not using setOperationAction.

**Comment by bjope** [comment]
> > A target could just promote the input vector elements itself if that is feasible when doing CustomLowerNode the first time.
> 
> For non-MVT types, I don't think there's any way for a target to request custom lowering. At least, not using setOperationAction.

Right. CustomLowerNode would just bail out early for such types. I did not consider that, so thanks for enlightening me.
I'll take another look at your proposal related to "construct an EXTRACT_VECTOR_ELT with the promoted vector type".

**Comment by arsenm** [comment]
> For non-MVT types, I don't think there's any way for a target to request custom lowering. At least, not using setOperationAction.

MVT::Other is supposed to work for this

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 1]
```diff
@@ -3443,11 +3438,13 @@ SDValue DAGTypeLegalizer::SplitVecOp_EXTRACT_VECTOR_ELT(SDNode *N) {
...
-    EltVT = MVT::i8;
-    VecVT = EVT::getVectorVT(*DAG.getContext(), EltVT,
-                             VecVT.getVectorElementCount());
+  if (!EltVT.isByteSized()) {
+    EltVT = EltVT.changeTypeToInteger().getRoundIntegerType(*DAG.getContext());
+    VecVT = VecVT.changeElementType(EltVT);
     Vec = DAG.getNode(ISD::ANY_EXTEND, dl, VecVT, Vec);
+    SDValue NewExtract =
+        DAG.getNode(ISD::EXTRACT_VECTOR_ELT, dl, EltVT, Vec, Idx);
+    return DAG.getZExtOrTrunc(NewExtract, dl, N->getValueType(0));
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:3447]: Why isn't this any ext or trunc?

Comment (bjope) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp]: Nice catch. I've updated it to use getAnyExtOrTrunc now.

**Comment by efriedma-quic** [comment]
Is there some reason we want to create a new EXTRACT_VECTOR_ELT, but not a new INSERT_VECTOR_ELT?

**Comment by bjope** [comment]
> Is there some reason we want to create a new EXTRACT_VECTOR_ELT, but not a new INSERT_VECTOR_ELT?

I think that some earlier step should have choosen to widen instead of split the vector if such legalization is preferred for the INSERT_VECTOR_ELT case.
And since SplitVecRes_INSERT_VECTOR_ELT should return a pair of vectors (the lo/hi halfs) it is a bit unclear to me how to do it.. One widened INSERT_VECTOR_ELT with a split+trunc afterwards? And then hope for that the split+trunc are cheap to legalize.

I think optimizing that situation is out-of-scope for this patch anyway. This patch is just supposed to support some scenatios for which we either hit assertion failure of miscompiled in the past. But if suddenly the legalization here is considered as performance critical, then we can follow up with optimizations later. We would need some test cases that show the benefit of doing it some other way, and I don't know about any such test cases.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Review by arsenm — APPROVED** [review]

**Pull request merged**